compiler/src/main/java/org/qbicc/type/definition/element/InitializerElement.java

```diff
@@ -48,14 +49,45 @@ public MethodBody getOrCreateMethodBody() {
                 synchronized (this) {
                     methodBody = this.methodBody;
                     if (methodBody == null) {
-                        this.methodBody = methodBody = factory.createMethodBody(methodBodyFactoryIndex, this);
+                        if (inProgress) {
+                            throw new IllegalStateException("Recursive method body creation");
```

Comment:
This may be too restrictive given that mutually recursive initializers are valid (though can lead to deadlocks).

Is this to defend against the mutually recursive case? or against multi-threaded processing of the initializer? If it's the later, we'll need to implement more of the class initializion process at compile time to make the second attempt wait for the first

Reply:
This method should only ever be called from `Driver` or in cases where we know that recursion isn't possible.  The  `tryCreateMethodBody` method covers the cycle case by returning when one is detected.  The key is that the *compilation* of a class initializer will be complete before the *compilation* of any access to its members, barring the cyclic case (in which some of one or more initializers might not be completely done initializing before the rest of them are).  In this way it mirrors how class initialization itself works, and the state machine is essentially similar, and I don't think it's really possible to do better than that without a dramatic increase in complexity.

Like class initialization, a deadlock is a very real possibility if two classes reference constants from each others' class in their initializer in the case that the two classes are simultaneously initialized from different threads.  But like the class initialization case, I think we don't have to care too much about that and can just let it fail.

Reply:
> This method should only ever be called from `Driver` or in cases where we know that recursion isn't possible.

Is there some documentation (javadoc?) that indicates this contract?  It would be good to get that recorded somewhere.

> The key is that the _compilation_ of a class initializer will be complete before the _compilation_ of any access to its members, barring the cyclic case (in which some of one or more initializers might not be completely done initializing before the rest of them are).

I'm still unclear on why this is a requirement for most class initializers, apart from the benefits of peephole optimizing away the `getstatic` to drive early DCE.  The critical case in my mind is the `isWindows`-style constants  and, as suggested on the call, I think special casing those native probe classes to occur first makes a lot of the complexity in the rest of this code drop away.

> In this way it mirrors how class initialization itself works, and the state machine is essentially similar,

+1 and to not caring about the deadlock case, as that can be flagged at build time.

> I don't think it's really possible to do better than that without a dramatic increase in complexity.

I think there's less complexity in special casing the few native probe classes and their constants - as they are clearly special - and allowing the rest of this code to sort of fall out in the wash.  Otherwise, we'll be transitioning back and forth between the compiler and the interpreter to get initializers processed and drive the additional class loads.

Reply:
> > This method should only ever be called from `Driver` or in cases where we know that recursion isn't possible.
> 
> Is there some documentation (javadoc?) that indicates this contract? It would be good to get that recorded somewhere.

Yeah that's a good idea. I'll add some docs for this.

> > The key is that the _compilation_ of a class initializer will be complete before the _compilation_ of any access to its members, barring the cyclic case (in which some of one or more initializers might not be completely done initializing before the rest of them are).
> 
> I'm still unclear on why this is a requirement for most class initializers, apart from the benefits of peephole optimizing away the `getstatic` to drive early DCE. The critical case in my mind is the `isWindows`-style constants and, as suggested on the call, I think special casing those native probe classes to occur first makes a lot of the complexity in the rest of this code drop away.

It's all about early DCE. The more constant conditions we can detect, the faster we can compile (and of course the more chances we have for later DCE, since we don't iterate through phases more than one time).

I'm not really sure a special case would help here. There will always be another case which benefits from early DCE (first it's platform, then it could be CPU type, then it might be page size, or a constant array size value, or magic numbers like bitmasks, or even complex things like build-time application configuration such as Quarkus uses, etc.).  If you extend a special case to enough situations, eventually it becomes a general case.

> > I don't think it's really possible to do better than that without a dramatic increase in complexity.
> 
> I think there's less complexity in special casing the few native probe classes and their constants - as they are clearly special - and allowing the rest of this code to sort of fall out in the wash. Otherwise, we'll be transitioning back and forth between the compiler and the interpreter to get initializers processed and drive the additional class loads.

I'm not really sure that it'd be less complex in the end. I think it's *generally* easier to maintain these kinds of things as source code than as intrinsics or any other early replacement programmatic mechanism. It seems to me that we'll have to transition back and forth between the compiler and interpreter anyway if we support any kind of build-time execution, since execution of class initialization can cause new classes to be loaded and can call any method which can also load classes, call things reflectively, or even generate classes dynamically to load. I don't believe it would be possible (or advisable) to ensure that all possibly reachable code is pre-compiled. So at least we're already going to be talking about waiting for the compiler to allow initialization to proceed. And if we're supporting constant folded methods like GraalVM, then the compiler will have to execute the method in question during compilation already, so there's another case where the compiler should be able to call into the interpreter. I'm not sure we can reasonably separate these worlds forever.